File: lc/coreprep/implementation/practice_iterations/patterns_iterations_2/NextLetter.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class NextLetter {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();
      in.nextLine();

      while (tt-- > 0) {
        String s = in.next();
        char k = in.next().charAt(0);

        System.out.println(nextLetter(s, 0, s.length() - 1, k));
      }
    }
  }

  // O(logn) / O(1)
  public static char nextLetter(String s, int low, int high, char k) {
    if (k < s.charAt(low) || k > s.charAt(high)) {
      return s.charAt(low);
    }

    while (low <= high) {
      int mid = low + (high - low) / 2;

      if (s.charAt(mid) <= k) {
        low = mid + 1;
      } else if (s.charAt(mid) > k) {
        high = mid - 1;
      }
    }

    return s.charAt(low % s.length());
  }
}

File: lc/coreprep/implementation/practice_iterations/iteration_archives/modified_binary_search_1/next_letter.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

// O(logn) / O(1)
char next_letter(const string &s, char k) {
  if (k < s[0] || k == s[s.size() - 1]) {
    return s[0];
  }

  int low = 0, high = s.size() - 1;
  while (low <= high) {
    int mid = low + (high - low) / 2;

    if (s[mid] <= k) {
      low = mid + 1;
    } else {
      high = mid - 1;
    }
  }

  return s[low % s.size()];
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt;
  string s;
  char k;

  cin >> tt;
  cin.ignore(1, '\n');

  while (tt--) {
    cin >> s >> k;
    cout << next_letter(s, k) << "\n";
  }

  return 0;
}